### ISSUE TITLE
GDN-hybrid models (Qwen3.8-27B) produce garbage tokens on RTX 5080 (sm_120a) — dense models fine; pin b81c99b suspected

### ISSUE THREAD

**Issue by zack0418** [kvmem/kvmem-llama.cpp#1]
# GDN-hybrid models (Qwen3.8-27B) produce garbage tokens on RTX 5080 (sm_120a) — dense models fine; pin b81c99b suspected

## Summary

KVMem v0.14.0 built per README (WSL2 Ubuntu 22.04.5, CUDA 13.2, `build-cuda.sh` defaults, arch `120a-real`) on an **RTX 5080 16 GB** starts fine and reaches `{"status":"ok"}`, but **text generation is incoherent token soup** for Qwen3.8-27B GGUFs, at `temperature 0` greedy. A dense model (Qwen3-VL-8B-Instruct Q4_K_M) on the same build answers correctly, so the engine, CUDA stack, and serving path are healthy — the failure tracks the **Qwen3.8 GDN-hybrid model path** (both a K-quant mix and an i-quant IQ3_S fail), not KVMem itself.

## Repro

1. Build v0.14.0 per README on sm_120a.
2. Serve any Qwen3.8-27B GDN-hybrid GGUF, e.g.:

```bash
./build/bin/llama-kvmem-server \
  -m Qwen3.8-27B-GSQ-RCO-IQ3_S-mtp.gguf \
  -c 262144 -n 16384 -b 512 -ngl 99 \
  --kvmem --kvmem-method retrieval \
  --kvmem-query-replay auto --kvmem-query-policy user \
  --kvmem-budget 36864 --kvmem-gen-reserve 16384 \
  --kvmem-block-tokens 128 --kv-dtype q8_0 \
  --spec-type draft-mtp --spec-draft-n-max 2 --spec-kv-dtype f16 \
  --enable-thinking --reasoning-budget 4096 \
  --host 127.0.0.1 --port 18200
```

3. `POST /v1/chat/completions` with "What is 2+3? Answer with the number only.", `temperature 0`, `max_tokens 32` → HTTP 500 (`chat_out_parse_fail … unparsed peg-native output`) with gibberish content; sample (greedy):

```
<n on  ch佑唯 s\n\n  basatan钧 Spi汇报 to onless Huge processing bras  5 anartchi alikeiam
```

MTP acceptance is 0.0% (`KVMEM_TRACE spec_stats n_drafted=63 n_accept=0`), consistent with broken target logits.

## Discrimination matrix (all on the same build/GPU)

| # | Variant | Result |
|---|---|---|
| A | `--spec-type none` | still garbage → speculative layer exonerated |
| B1 | KVMem off entirely (`f16` KV, 32K ctx, plain path) | still garbage → **KVMem layer and quantized KV exonerated** |
| C | Qwen3-VL-8B-Instruct Q4_K_M (dense) | **correct** ("5") → build/CUDA/5080 healthy |
| D | Ridge-3.7bpw (Qwen3.8 GDN hybrid, K-quants, no i-quants) | still garbage → quant family exonerated |

Also verified: model files are bit-identical to sources that run correctly on llama.cpp builds dated ~2026-09-10 on the same GPU (Windows fork), so the GGUFs are fine — the delta is the **pinned llama.cpp (b81c99b, 2026-09-02)**.

## Environment

- WSL2 (Ubuntu 22.04.5), RTX 5080 16 GB, driver 616.92 / WSL CUDA UMD 13.4, toolkit 13.2 (V13.2.51)
- kvmem v0.14.0, submodule pin `b81c99b479d4c24e5eeca10de99032ebd343ef8f` (2026-09-02), built `-DGGML_CUDA_FA_ALL_QUANTS=ON`, arch `compute_120a`

## Patch porting data (may help a rebase)

`patches/llama-kvmem-current.patch` vs upstream **v0.4.1** (b29c606): 18/21 files apply cleanly (`git apply --3way --stat`); 3 files conflict:

- `common/speculative.cpp` — 29-line .rej (3× `batch.logical_pos` assignment sites)
- `src/models/qwen35.cpp` — 23-line .rej (2× `kvmem_capture_q/k/v` hook sites)
- `tools/server/server-common.cpp` — 61-line .rej (`oaicompat_chat_params_parse` was refactored upstream into `oaicompat_chat_process_media`)

Note the README results were measured on a 5060 Ti (also sm_120a) — if you've seen this model family work there with this pin, the dispatch difference by SM count may matter; on the 5080 both GDN-hybrid quants fail while a dense model passes on the identical build.

Happy to run any additional probes. Thanks for KVMem!

**Comment by wangleye** [comment]
Thank you. I'll check how to reproduce your results first.

**Comment by wangleye** [comment]
@zack0418 I cannot reproduce the error. But I just built a pre-release version on the current version. If possible, could you help to try this build and see whether the error sttill exists.

**Comment by zack0418** [comment]
@wangleye Thanks for the quick turnaround! Two updates from the RTX 5080 (sm_120a, GB203, WSL2):

**1. The garbage output is gone in v0.16.0-rc1** ✅

Same model (GSQ-RCO IQ3_S-mtp), same recipe, greedy temp-0: correct answer, clean reasoning, `chat_out_parse_fail` = 0, 200 OK.

One note: master still pins llama.cpp at b81c99b, and our locally-built v0.14.0 (system CUDA 13.2.51 toolkit) produced garbage on this GPU while your 5060 Ti validation was fine. If v0.16.0-rc1 didn't change the pin either, the difference may be toolchain-side (your BUILD-INFO shows a private nvcc env; our build used distro CUDA 13.2) — i.e., possibly a compiler-version-dependent miscompilation on sm_120a/GB203 rather than a source bug. Might be worth noting which nvcc the release was built with.

**2. New observation — retrieval recall on dense single-document audit**

Setup: IQ3 recipe (36K budget / 16K reserve), one 34K-char scientific paper buried at ~90% depth behind 203K chars of unrelated filler; prompt = 106,725 tokens. Then five detail questions about the buried paper (TGA heating rate + OCR artifact, author affiliation, a specific 10 Hz current sweep, an electrode radius stated as "500 mm" (μ lost), and an abstract-vs-conclusions contradiction).

Result:

- 3 of 5 answered "not mentioned in the text" — including two where the exact figures (10 Hz: 1→1.45 mA read at 0.2 V; Ta radius 500 μm) are stated verbatim in the buried paper
- affiliation answered as a different institution (possibly retrieved from filler context)
- no hallucination anywhere — honest omissions, which is the right behavior for what the model could see
- the same model at full attention (small context) answers all five perfectly

So retrieval is silently dropping audit-critical blocks on dense single-document QA — no errors, just missing evidence. Understandable given the agent-memory focus of the paper (LongMemEval/AgentLongBench are multi-session QA), but it means dense single-doc audit likely needs a bigger retrieval budget or query expansion. Happy to run more probes (e.g., varying `--kvmem-budget`, needle depth, or filler type) if that helps.

Environment: WSL2 Ubuntu 22.04.5, prebuilt `linux-x86_64-cuda13-sm120a` tarball (sha256 verified), RTX 5080 16 GB, driver 616.92.

**Comment by wangleye** [comment]
@zack0418 Thank you for the feedback. Can you give me the paper title as well as the questions? I will also check myself.

Besides, some parameter tuning may help. For long-context truth finding, our experiment finds small block size is better. The default block size now is 128, but you can try 32, i.e., adding this to the script "--kvmem-block-tokens 32".

**Comment by zack0418** [comment]
@wangleye 谢谢！信息如下：

**论文**：《Synaptic plasticity and learning behaviours in flexible artificial synapse based on polymer/viologen system》— Chaochao Zhang, Yu-Tsung Tai, Jie Shang, Gang Liu, Kun-Li Wang, Chienwen Hsu, …, Run-Wei Li，J. Mater. Chem. C, 2016, 4, 3217–3223。针文档 = 论文全文 markdown（约 34K 字符），埋在 203K 字符无关填充后约 90% 深度，prompt 共 106,725 tokens。

**五个问题**（中文原样提问）：
1. TGA 实验的温度范围和升温速率是什么？文中数值有无 OCR 伪影？如有，真实值应是多少？
2. 该研究的作者机构全名是什么？
3. 10 Hz 刺激时电流从多少升到多少？读电压是多少？
4. Ta 顶电极的半径文中写的是多少？物理上合理值应为多少？
5. 作者结论中关于"弯曲 100 次"的说法与正文数据有无矛盾？具体说明。

**两种 block 的结果——数字召回问题都没过：**

- **block 128（默认）**：③"文中没有出现 10 Hz 刺激条件"、④"文中没有给出 Ta 顶电极半径"——而针文档里逐字写着"10 Hz 时电流从 1 mA 升至 1.45 mA（读电压 0.2 V）"和"radius of 500 mm"（PDF 转 μ 丢失，实为 500 μm）。②答了机构 a（宁波材料所，正确）并如实说明 b/c/d 缺失；⑤发现"结论证据不足"但漏了 r=8 mm 失效阈值。
- **block 32（你的建议）**：③同样失败——它正确引用了"1–10 Hz 扫描"原句（说明部分针块确实被检索到了）但说"没有具体数值"；④更糟——**编造了文中不存在的句子**（"a diameter of 2 mm"，needle 全文 "diameter" 出现 0 次，原文实为 "a radius of 500 mm"）；②把 klwang@ntut.edu.tw 推断成 "National Tsing Hua University"（实为 National Taipei University of Technology）。⑤类似半对。
- 稳定性：block 32 首次运行中途连接中断（客户端侧原因），重试完成。

**观察**：两种粒度下，埋入论文的部分块都能被检索到（模型能引用原文句子），但**承载具体数值的块（LTP 图值、制作半径、TGA 速率）两种粒度都丢**。block 128 是诚实省略，block 32 出现编造。同一模型全注意力（小上下文）下五问全对。

需要的话我可以继续跑更多探针（不同 --kvmem-budget、针深度、填充类型）。环境：WSL2 Ubuntu 22.04.5，v0.16.0-rc1 prebuilt linux-x86_64-cuda13-sm120a（sha256 已验），RTX 5080 16 GB，驱动 616.92。

**Comment by zack0418** [comment]
@wangleye 补充一组数据：按你的建议测了 block 32，同时把 `--kvmem-budget` 也做了阶梯（36K / 64K / 96K，gen_reserve 16384 不变，block 128 控制变量）。结论：**budget 越大，预填充速度坍塌越狠，36K 就是 16G 卡的实际上限**。

同一份 106,725 tokens 的深针 prompt（203K 字符填充 + 34K 字符论文 @90% 深度）：

| --kvmem-budget | 预填充吞吐（512 行/chunk 实测） | 状态 |
|---|---|---|
| 36,864 | ~220 tok/s | 完整跑完（检索召回问题见上） |
| 65,536 | ~100 tok/s | 预填至 101K/106K 时客户端 850s 超时 |
| 98,304 | ~29 tok/s | 预填至 36K/106K，预计需数小时，放弃 |

服务器本身不崩，就是慢——预算增大后块调度/检索簿记的开销在预填阶段非线性增长。结合上面的深针结果（36K 下数字召回已不合格），在 16G 卡上"加大预算换召回"这条路走不通；除非后续版本能修复预算-预填的伸缩性，否则密排单文档审计场景我们只能回到全注意力或无损方案。

block 32 的召回结果（两种粒度都没过数字题，且 32 出现一处编造）已包含在上面，供你自查参考。需要完整输出（block 128/32 深针 + 五题）随时说。

**Comment by wangleye** [comment]
好，我去测一下

**Comment by wangleye** [comment]
@zack0418 另外cuda构建的问题我彻底搞清楚了，需要13.2.86以上版本（之前的版本对iq3量化有bug）
